### DISCUSSION TITLE
Negative binomial abundance estimates for temporal model

### DISCUSSION THREAD

**Discussion by naph0n** [inlabru-org/inlabru#167] [Q&A] [answered]
Hi all,

I'm currently working on a project that requires me to obtain abundance estimates of various marine mammal species over certain time periods. I have fitted a negative binomial model for the counts and am struggling to find the mean predictions and abundance estimates for the different time points (e.g., predictions for each season). 

I have tried the predictions in the same way as for a LGCP but I understand that it needs to be done differently in this case. I know that I could sample the size parameter and the probabilities and generate counts to predict abundance but have had no success so far.

Does anyone have any ideas? Let me know if you need to see any code, etc.

Thanks!

**Comment by finnlindgren** [comment, marked as answer]
Are you trying to do predictions at individual locations, or some kind of spatial aggregation?

You can use `generate(fit, n.samples = 1)` and look at the output to see which variables and parameters will be available to use in a prediction expression (that's the easiest way to find out what the parameter names will be in the output).  Then if you want the probability mass function at a location, you'll need something like this:
```{r}
K <- 0:max_K
pred <- predict(fit, data = newdata, formula = ~ dnbinom(K, size = ... , mu = E * exp(...)))
```
where you replace the "..." with the needed parameter name and linear predictor expression. 
It's a bit messier to do that for multiple locations at the same time, but can be done.

If you just want the posterior mean and variance,
```{r}
pred <- predict(fit, data = newdata, formula =
  ~ {
       mu <- E * exp(your predictor expression)
       list(m = mu,
         sigma2 = mu * (1 + mu / name_of_size_parameter))
    }
  )
posterior_mean <- pred$m$mean
posterior_variance <- pred$sigma2$mean + pred$m$sd^2
``` 
I picked the `sigma2` formula from `INLA::inla.doc("nbinomial")`. The posterior predictive variance is $V(Y | data) = E_\theta[V(Y | \theta, data) | data] + V_\theta[E(Y | \theta, data) | data] = E(\sigma^2 | data) + Var(\mu | data)$.

**Reply by naph0n** [reply]
Thank you for your response! I am trying to do predictions at individual locations just now. 

What would I define `E` and `newdata` as? I defined `E` as `E = runif(n, min, max)` like in the documentation but got this error: 

`In E * exp(mySmooth + myOffset + Intercept) : longer object length is not a multiple of shorter object length`

**Reply by finnlindgren** [reply]
`E` should be your "exposure" or "effort" scaling information. If you don't use that in your model, it should be all ones (or simply left out of the computations).  Which documentation do you mean? If it's our documentation, that `E` construction is probably just to construct a synthetic data set. We may need to clarify that, so please let me know precisely where it is...

**Reply by naph0n** [reply]
I got `E` from the example in `INLA::inla.doc("nbinomial")`!

**Reply by finnlindgren** [reply]
I see. The part
```r
n = 1000
x = rnorm(n, sd = 0.2)
eta = 1 + x
E = runif(n, min = 0, max=10)
mu = E * exp(eta)
size = 3
y = rnbinom(n, size=size, mu=mu)
```
is purely there to generate a synthetic data set. When you have real data you need to use that instead.

**Reply by naph0n** [reply]
That makes sense! I have changed `E` to my "effort" scaling information, which is the area of each point. However, for the probability mass function, I still get this warning, along with some strange results:

`In E * exp(mySmooth + myOffset + Intercept) : longer object length is not a multiple of shorter object length`

Maybe there is an issue with the `newdata` that I am inputting into the predict call, but I'm not sure.

**Reply by finnlindgren** [reply]
That is most likely due to the details of what you're placing in the input data, and how you've written the expression to be evaluated. If you have more than a single row in your input data, you need a more complex expression than my example for `dnbinom`.

**Reply by finnlindgren** [reply]
To setup the nbinom calculation so that it calculates the entire function for each prediction point, use this technique, for example:
```r
newdata <- data.frame(a=1:3,b=2:4) # placeholder for example
library(tidyverse)
pred_data <- full_join(newdata %>% mutate(row=seq_len(n())),data.frame(K=0:5),by=character())
   a b row K
1  1 2   1 0
2  1 2   1 1
3  1 2   1 2
4  1 2   1 3
5  1 2   1 4
6  1 2   1 5
7  2 3   2 0
8  2 3   2 1
9  2 3   2 2
10 2 3   2 3
11 2 3   2 4
12 2 3   2 5
13 3 4   3 0
14 3 4   3 1
15 3 4   3 2
16 3 4   3 3
17 3 4   3 4
18 3 4   3 5
```
then you can use the example code.  Adding the row index is just to help identify each function after the predict call.

**Reply by finnlindgren** [reply]
See here for some related examples https://inlabru-org.github.io/inlabru/articles/prediction_scores.html; specifically, for how to compute prediction assessment scores. For log-prediction-probability scores you only need the dnbinom function at a single value, so can just replace `K` by the response variable name.

**Reply by naph0n** [reply]
I see, I understand that, just confused about what exactly to put in `newdata` - is it all the covariates from my main data frame?

**Reply by finnlindgren** [reply]
Yes, for prediction you typically need everything that you had in your original data, that you used in the model definition and model estimation. Except of course the actual measured value.  In other words, newdata should have the same structure as the original data, but "new".

**Reply by finnlindgren** [reply]
My answers assume you're predicting at new locations. If you want to compute estimates of the estimated model values at your original data locations, you can simply provide the original data object.

**Reply by naph0n** [reply]
Oh right! I just need estimates at the original locations, so will just use my main data set. Is it possible to also display the predictions over a pixels data frame of the boundary instead of the points?

**Reply by finnlindgren** [reply]
Yes. I think the code might automatically handle that if you supply the input as a SpatialPixelsDataFrame, but otherwise you can reformat the results into that afterwards.  The `gg()` methods in inlabru are meant to help with basic plotting, but in general that's really a job for other packages.

**Reply by naph0n** [reply]
That's great - thank you for all your help!
